### DISCUSSION TITLE
How to get all pushed commit information

### DISCUSSION THREAD

**Discussion by nackdai** [community/community#25797] [Actions]
<p>I would like to know how to get information of all commits which I pushed.</p>
<p>For example, if I pushed 2 commits like below, can I get information of all pushed commits?</p>
<pre><code class="lang-auto">commit xxxxxx
Author: xxxxx &lt;xxxx@xxxx&gt;
Date:   Sat Jun 6 xx:xx:xx 2020 +0900

    Update

commit yyyyy
Author: xxxxx &lt;xxxx@xxxx&gt;
Date:   Fri Jun 5 xx:xx:xx 2020 +0900

    Add
</code></pre>
<p>Especially, I would like to get changed files in all pushed commits.</p>

**Comment by ghost** [comment]
<p>If you’re using <code>actions/checkout@v2</code> <a href="https://github.com/actions/checkout/">https://github.com/actions/checkout/</a>,  it fetches only a single commit by default.</p>
<p>Would setting <code> fetch-depth: 0</code> to fetch all history for all tags and branches help with what you are trying to achieve?<br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches" target="_blank">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/10137/181093326-bad88070-30dc-4131-9360-8457db237612.png" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches" target="_blank">actions/checkout</a></h3>

<p>Action for checking out a repo. Contribute to actions/checkout development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by nackdai** [comment]
<p>Thank you for your response.</p>
<p>Can we get how many commits we did?<br>
In my understanding, even if we have all repository logs, if we don’t know how many commits we push, we don’t have any way to get modified files name which we push.</p>
<p>Just what I would like to know is modified files name which we push.<br>
In this case, “push” doesn’t mean all history, that means one push I did at some point.</p>
<p>I don’t mind to use Dockerfile or JavaScript for realizing it.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/nackdai">@nackdai</a>,<br>
Sorry for the delay response.<br>
In push event, <code>before</code> and <code>after</code> key are stand for the previous commit of this push operation and the last commit in this push operation. All the commits related to the push are included between these two commits.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/6a015ff53130e7a3f814524d6ec9a1c7f099179c.png" data-download-href="/uploads/short-url/f7LyyN7Vbz4bnSRrvH5P1C7bObi.png?dl=1" title="pr event" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181093328-3353a497-872c-4cea-a664-d18737fff570.png" alt="pr event" data-base62-sha1="f7LyyN7Vbz4bnSRrvH5P1C7bObi" width="561" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/6a015ff53130e7a3f814524d6ec9a1c7f099179c_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">pr event</span><span class="informations">609×542 13.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>To get the modified files accossiated with the push event, you could use <code>git diff</code> command to get the file names.<br>
<code> git diff --name-only ${{ github.event.before }} ${{ github.event.after }}</code></p>
<p>Also, remember to set fetch-depth: 0 in <code>checkout</code> action.<br>
Please see my example workflow yml :</p>
<pre><code class="lang-auto">on:
  push:
jobs:
   build: 
     runs-on: ubuntu-latest
     steps: 
      - uses: actions/checkout@v2
        with: 
          fetch-depth: 0
      - name: get file changes
        run: | 
          git diff --name-only ${{ github.event.before }} ${{ github.event.after }}
</code></pre>

**Reply by matcaampos** [reply]
Hi man, is there a way to know how many commits are in a push?

**Reply by batyrf** [reply]
`fetch-depth` should be set to `0`

`git log --pretty=oneline "$BEFORE_REF...$AFTER_REF" | wc -l`.

**Reply by yolicap** [reply]
@matcaampos This stackoverflow article gives a much cleaner answer than the one suggest previously
https://stackoverflow.com/questions/31997999/number-of-commits-between-two-commitishes

Basically: `git rev-list --count A..B`

**Comment by rodriwito** [comment]
<p>I have some questions.<br>
How could I get the commits between the <em>before</em> and <em>after</em>?<br>
What happen if the <em>before</em> is a commit that in the push has changed for example by rebase, and after push have totally different key?</p>

**Reply by batyrf** [reply]
You can get commits in between before and after with this:
```
- name: log newly committed messages
  env:
    BEFORE_REF: ${{ github.event.before }}
    AFTER_REF: ${{ github.event.after }}
  run: git log --pretty=format:%s "$BEFORE_REF...$AFTER_REF"
```

⚠️ `fetch-depth` should be set to `0`
